p2panda-spaces/src/space.rs

```diff
@@ -374,6 +490,89 @@ where
         Ok(events)
     }
 
+    /// Process an auth control message before the "authored" version has been forged. This is
+    /// useful when we want to know what the resulting auth state will be.
+    async fn pre_process_auth(
+        manager: Manager<S, F, M, C, RS>,
+        space_id: ActorId,
+        dependencies: &Vec<OperationId>,
+        control_message: &AuthControlMessage<C>,
+    ) -> Result<(Vec<ActorId>, Vec<ActorId>), SpaceError<S, F, M, C, RS>> {
+        let manager = manager.inner.read().await;
+        let my_id: ActorId = manager.forge.public_key().into();
+        let auth_y = manager.store.auth().await.map_err(SpaceError::AuthStore)?;
+        let previous_members = secret_members(auth_y.members(space_id));
+
+        // We process a fake operation on the current auth state in order to compute the next
+        // membership state.
+        let fake_op = AuthMessage::Forged {
+            author: my_id,
+            operation_id: OperationId::placeholder(),
+            args: AuthArgs {
+                dependencies: dependencies.to_owned(),
+                control_message: control_message.to_owned(),
+            },
+        };
+
+        // NOTE: as we are only calling this method when we ourselves want to make space
+        // membership changes locally, no operations will be concurrent to our local auth
+        // state, and so no re-build will occur. This means processing the operation is cheap.
+        // Processing the operation through the auth api is preferred over calculating the
+        // next state manually, as auth does as some basic validation of control messages for
+        // us.
+        let auth_y_i = AuthGroup::process(auth_y, &fake_op).map_err(SpaceError::AuthGroup)?;
+        let next_members = secret_members(auth_y_i.members(space_id));
+        let diff_members = next_members
+            .iter()
+            .cloned()
+            .filter(|actor| previous_members.contains(actor))
+            .collect::<Vec<_>>();
+        Ok((next_members, diff_members))
+    }
+
+    fn update_orderer_states(
+        mut encryption_y: EncryptionOrdererState,
+        mut auth_y: AuthOrdererState,
+        message: &M,
+    ) -> (EncryptionOrdererState, AuthOrdererState) {
+        match message.args() {
+            SpacesArgs::KeyBundle {} => unimplemented!(),
+            SpacesArgs::ControlMessage {
+                auth_dependencies,
+                encryption_dependencies,
+                ..
+            } => {
+                auth_y.add_dependency(message.id(), auth_dependencies);
+                encryption_y.add_dependency(message.id(), encryption_dependencies);
+            }
+            // @TODO: also include application messages in auth and encryption dependencies.
+            SpacesArgs::Application { .. } => (),
+        };
+        (encryption_y, auth_y)
+    }
+
+    // Persist both auth and space state.
+    async fn persist_space(
```

Comment:
How about calling this `set_state`? Then it's nicely fitting the `state` method.

Reply:
It could also move closer to `state` for readability